ddtrace/tracer/spancontext.go

```diff
@@ -389,42 +420,52 @@ func (t *trace) finishedOne(s *span) {
 		// TODO(barbayar): make sure this doesn't happen in vain when switching to
 		// the new wire format. We won't need to set the tags on the first span
 		// in the chunk there.
-		for k, v := range t.tags {
-			s.setMeta(k, v)
-		}
-		for k, v := range t.propagatingTags {
-			s.setMeta(k, v)
-		}
-		for k, v := range ginternal.GetTracerGitMetadataTags() {
-			s.setMeta(k, v)
-		}
-		if s.context != nil && s.context.traceID.HasUpper() {
-			s.setMeta(keyTraceID128, s.context.traceID.UpperHex())
-		}
+		t.setTraceTags(s, tr)
 	}
-	if len(t.spans) == t.finished {
-		defer func() {
-			t.spans = nil
-			t.finished = 0 // important, because a buffer can be used for several flushes
-		}()
-	}
-	tr, ok := internal.GetGlobalTracer().(*tracer)
-	if !ok {
+
+	if len(t.spans) == t.finished { // perform a full flush of all spans
+		t.finishChunk(tr, &chunk{
+			spans:    t.spans,
+			willSend: decisionKeep == samplingDecision(atomic.LoadUint32((*uint32)(&t.samplingDecision))),
+		})
+		t.spans = nil
 		return
 	}
-	setPeerService(s, tr.config)
-	if len(t.spans) != t.finished {
-		return
+
+	doPartialFlush := tr.config.partialFlushEnabled && t.finished >= tr.config.partialFlushMinSpans
+	if !doPartialFlush {
+		return // The trace hasn't completed and partial flushing will not occur
 	}
-	if hn := tr.hostname(); hn != "" {
-		s.setMeta(keyTracerHostname, hn)
+	log.Debug("Partial flush triggered with %d finished spans", t.finished)
+	telemetry.GlobalClient.Count(telemetry.NamespaceTracers, "trace_partial_flush.count", 1, []string{"reason:large_trace"}, true)
+	finishedSpans := make([]*span, 0, t.finished)
+	leftoverSpans := make([]*span, 0, len(t.spans)-t.finished)
+	for _, s2 := range t.spans {
+		if s2.finished {
+			finishedSpans = append(finishedSpans, s2)
+		} else {
+			leftoverSpans = append(leftoverSpans, s2)
+		}
+	}
+	// TODO: (Support MetricKindDist) Re-enable these when we actually support `MetricKindDist`
+	//telemetry.GlobalClient.Record(telemetry.NamespaceTracers, telemetry.MetricKindDist, "trace_partial_flush.spans_closed", float64(len(finishedSpans)), nil, true)
+	//telemetry.GlobalClient.Record(telemetry.NamespaceTracers, telemetry.MetricKindDist, "trace_partial_flush.spans_remaining", float64(len(leftoverSpans)), nil, true)
+	finishedSpans[0].setMetric(keySamplingPriority, *t.priority)
+	if s != t.spans[0] {
+		// Make sure the first span in the chunk has the trace-level tags
+		t.setTraceTags(finishedSpans[0], tr)
 	}
-	// we have a tracer that can receive completed traces.
-	atomic.AddUint32(&tr.spansFinished, uint32(len(t.spans)))
-	tr.pushTrace(&finishedTrace{
-		spans:    t.spans,
+	t.finishChunk(tr, &chunk{
+		spans:    finishedSpans,
 		willSend: decisionKeep == samplingDecision(atomic.LoadUint32((*uint32)(&t.samplingDecision))),
 	})
+	t.spans = leftoverSpans
```

Comment:
It doesn't change with partial flushing, but at the moment that's intentional, at least to me.
I don't think the root should change if part of it is flushed, regardless of whether the root has been flushed or not. Earlier in this function, we lock down the priority once the root has finished, which means that it's possible that the priority of the trace could change even after some of the spans have been flushed, but there isn't really a way for us to avoid this right now.

LMK if you think there's a concern with the root in this case

Reply:
We're going to chat with ASM offline about this to make sure it's not a problem, and can rollback this PR if needed. (Or just indicate in the release notes that this feature may interact badly with ASM if turned on, while it's in beta)